### DISCUSSION TITLE
Using the keyboard for mouse navigation

### DISCUSSION THREAD

**Discussion by leandwo** [kmonad/kmonad#370] [General]
Currently, there isn't platform-agnostic way to provide some sort of mouse emulation in kmonad. 

https://github.com/kmonad/kmonad/issues/322 brought some mouse navigation projects to light which could pair well with kmonad to keep your hands on the keyboard. These were the ones mentioned so far: 

- https://github.com/tweekmonster/hammerspoon-vimouse
- https://github.com/unclechu/place-cursor-at
- https://github.com/jordansissel/keynav
- https://github.com/SalmonChris/KeyboardMouse
- https://github.com/Ph0enixKM/Mousepad
- https://www.semicomplete.com/projects/keynav/
- https://github.com/rvaiya/warpd
- the native mousekey feature in windows/macOS

Now that GH discussions are enabled in this repo, I'm extending the discussion here for visibility. If you come across any other ways to substitute mouse navigation with the keyboard, please do share! 🎃

**Comment by MuhammedZakir** [comment]
<i>@</i>david-janssen has said that he won't add this into KMonad, but will add hooks so one can implement this. For now, you can use `xdotool` (for Linux X11) and `clicclick` for Mac to generate mouse event when you press a key. Don't know about windows.

Edit: Just found this Python library - https://github.com/boppreh/mouse.

---

@david-janssen: IMHO, sending mouse events using a virtual mouse, just like we do for keyboards, will be better and will also be consistent. High-level APIs may not support all features on all platforms. For instance, Mac's `CGEvent` doesn't support sending a key down only event. Is this feasible? For Mac, IIAR, we will only need to change the code a little to support mouse (Karabiner dext supports mouse events). If we don't need much work on Linux and Windows sides too, it will be nice if this is supported.

**Reply by david-janssen** [reply]
I absolutely detest having to support Windows stuff, so anything that depends on expanding the OS-support for Windows is something I will struggle against with all my might. I'd be happy to hack on it in Linux, and admittedly, it looks really doable (just emitting kernel level pointer events). Find me a Windows programmer willing and able to take up the job, and I'm happy to give it a shot.

**Reply by molleweide** [reply]
> Edit: Just found this Python library - https://github.com/boppreh/mouse.

Hey Muhammed have you tried this with kmonad yet? I am looking at it now and it seems quite comprehensive.
I will try this in the comming days and try to report back.

Oh wait, I just realized that I wonder if the same problem will occur as in https://github.com/kmonad/kmonad/issues/426.
This will be interesting to find out.

**Reply by MuhammedZakir** [reply]
> Hey Muhammed have you tried this with kmonad yet? [...]

No.

**Reply by molleweide** [reply]
Now I have tested it (https://github.com/boppreh/mouse) and can confirm that it works on macos.

you have to install this with pip for it to work on macos:
```
	pyobjc-framework-Quartz                     # mouse dependency for macos
	git+https://github.com/boppreh/mouse.git    # generate mouse events with python
```

You have to create a script in which you create your mouse commands such as:
``` python
#!/usr/bin/env python3

import mouse

x, y = mouse.get_position()
mouse.move(x+50, y)
```

There is an apparent lag when trigger the cmd via key. I dunno how it will feel with a more built out set of commands.

**Comment by molleweide** [comment]
# A modal keyboard-driven virtual pointer

compatible with linux and macos from what I can see. written in C.

https://github.com/rvaiya/warpd
